lib/input/sftp.go

```diff
@@ -0,0 +1,495 @@
+	Codec                 string          `json:"codec" yaml:"codec"`
+	DeleteObjects         bool            `json:"delete_objects" yaml:"delete_objects"`
+}
+
+// SFTPCredentials contains the credentials for connecting to the SFTP server
+type SFTPCredentials struct {
+	Username string `json:"username" yaml:"username"`
+	Secret   string `json:"secret" yaml:"secret"`
+}
+
+// NewSFTPConfig creates a new SFTPConfig with default values.
+func NewSFTPConfig() SFTPConfig {
+	return SFTPConfig{
+		Server:                "",
+		Port:                  0,
+		Filename:              "",
+		Credentials:           SFTPCredentials{},
+		MaxConnectionAttempts: 10,
+		RetrySleepDuration:    "5s",
+		Path:                  "",
+		Codec:                 "lines",
+		DeleteObjects:         false,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// SFTP is a benthos reader.Type implementation that reads messages
+// from file(s) on an SFTP server.
+type SFTP struct {
+	conf SFTPConfig
+
+	log   log.Modular
+	stats metrics.Type
+
+	client *sftp.Client
+
+	objectScannerCtor codec.ReaderConstructor
+	keyReader         *sftpTargetReader
+
+	objectMut sync.Mutex
+	object    *sftpPendingObject
+}
+
+// NewSFTP creates a new SFTP input type.
+func NewSFTP(conf SFTPConfig, log log.Modular, stats metrics.Type) (*SFTP, error) {
+	var err error
+	var objectScannerCtor codec.ReaderConstructor
+	if objectScannerCtor, err = codec.GetReader(conf.Codec, codec.NewReaderConfig()); err != nil {
+		return nil, fmt.Errorf("invalid sftp codec: %v", err)
+	}
+
+	s := &SFTP{
+		conf:              conf,
+		log:               log,
+		stats:             stats,
+		objectScannerCtor: objectScannerCtor,
+	}
+
+	err = s.initSFTPConnection()
+
+	return s, err
+}
+
+type sftpObjectTarget struct {
+	key   string
+	ackFn func(context.Context, error) error
+}
+
+// ConnectWithContext attempts to establish a connection to the target SFTP server.
+func (s *SFTP) ConnectWithContext(ctx context.Context) error {
+	var err error
+	s.keyReader, err = newSFTPTargetReader(ctx, s.conf, s.log, s.client)
+	return err
+}
+
+func deleteSFTPObjectAckFn(
+	client *sftp.Client,
+	key string,
+	delete bool,
+	prev codec.ReaderAckFn,
+) codec.ReaderAckFn {
+	return func(ctx context.Context, err error) error {
+		if prev != nil {
+			if aerr := prev(ctx, err); aerr != nil {
+				return aerr
+			}
+		}
+		if !delete || err != nil {
+			return nil
+		}
+
+		_, err = client.Stat(key)
+		if err != nil {
+			return nil
+		}
+
+		aerr := client.Remove(key)
+
+		return aerr
+	}
+}
+
+//------------------------------------------------------------------------------
+
+type sftpTargetReader struct {
+	pending []*sftpObjectTarget
+	conf    SFTPConfig
+}
+
+type sftpPendingObject struct {
+	target    *sftpObjectTarget
+	obj       *sftp.File
+	extracted int
+	scanner   codec.Reader
+}
+
+func newSFTPTargetReader(
```

Comment:
Looks like the inspiration for this was taken from the s3 input, but it might be worth checking out the file input as well: https://github.com/Jeffail/benthos/blob/master/lib/input/file.go as it's simpler.

The reason for the complexity in the s3 version, and needing a generic object reader type, is because we need to support multiple strategies for obtaining a list of files to download, whereas here it looks like we're always walking a static list of files so we can probably reduce a lot of the code here.